opcua/server/address_space.py

```diff
@@ -180,34 +180,46 @@ def add_nodes(self, addnodeitems, user=User.Admin):
     def _add_node(self, item, user):
         result = ua.AddNodesResult()
 
-        if item.RequestedNewNodeId in self._aspace:
+        # Check if the null NodeId is given
+        if item.RequestedNewNodeId.is_null():
+            self.logger.debug("RequestedNewNodeId is null")
+            if item.BrowseName.NamespaceIndex:
```

Comment:
This is wrong all nodes should have a a parent, but OK, write a FIXME there
and leave it as is. But can you use the idx of parent instead of browse
name if nodeid is null? and remove the _default_namespace variable, we
should not have such a thing. If no parent and and null nodeid, we refuse
to create...

On Wed, 23 Mar 2016 at 09:40 Denis Štogl <notifications@github.com> wrote:

> In opcua/server/address_space.py
> <https://github.com/FreeOpcUa/python-opcua/pull/149#discussion_r57124528>:
>
> > @@ -180,34 +180,46 @@ def add_nodes(self, addnodeitems, user=User.Admin):
> >      def _add_node(self, item, user):
> >          result = ua.AddNodesResult()
> >
> > -        if item.RequestedNewNodeId in self._aspace:
> > +        # Check if the null NodeId is given
> > +        if item.RequestedNewNodeId.is_null():
> > +            self.logger.debug("RequestedNewNodeId is null")
> > +            if item.BrowseName.NamespaceIndex:
>
> But if we do that, than a lot of tests will be broken, and you can not add
> child in the new NameSpace.
>
> Also we can not return hard BadParentNodeIDInvalid if the parent is null,
> I tested it already, but there is a lot of server nodes which does not have
> parent specified so you can not start server any more...
>
> I agree we should not allow it from the Client, but at this level we have
> no idea if the server or client are creating nodes. On the other hand this
> could be solved using username variable, where we can define e.g.
> ServerAdmin. I would propose to merge this and open new Issue regarding
> this and I am willing to propose me as responsible for issue and than I
> create proposal for this in the next weeks....
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly or view it on GitHub
> <https://github.com/FreeOpcUa/python-opcua/pull/149/files/a6be8b95b418d4abec719247991a80265bbcec86#r57124528>
>